### ISSUE TITLE
Defer PreferencesService to lazy/later and/or worker thread

### ISSUE THREAD

**Issue by breedx-splk** [open-telemetry/opentelemetry-android#149]
As part of the first disk buffering PR #145, it was raised by @marandaneto that  the `PreferencesService` is loading shared preferences from the context, which does perform IO and will result in a warning when StrictMode is used. This suggests that this work should either be deferred until needed or moved to a worker thread.

See: 
https://github.com/open-telemetry/opentelemetry-android/pull/145#discussion_r1392352275

**Comment by LikeTheSalad** [comment]
I think the solution should be to run all the initialization in a background thread and provide callbacks for the user to know when OTel is ready to be used.

The reason I mention this is because `PreferencesService` isn't the only blocking work needed during initialization. The calls for [checking available cache disk space](https://github.com/open-telemetry/opentelemetry-android/blob/50206e59f9eb7dccbf6b49d46b929f7ecdbd5c26/instrumentation/src/main/java/io/opentelemetry/android/internal/services/CacheStorageService.java#L46) should also be done out of the main thread, and it'll be needed to create signals exporters for initializing the OpenTelemetry instance. Plus (probably won't be needed over here) though at least in the Elastic agent, we make a request to an NTP server during initialization to get the real time for signals' timestamps, since mobile devices' clock isn't too reliable for accuracy. There might be other solutions that wouldn't involve an NTP server though, so that's why I'm not sure if it'll be added here, however, my point is that initializing OTel in the main thread seems like it could cause troubles in the long run if we need to add features that take some time to set up.

**Comment by marandaneto** [comment]
> I think the solution should be to run all the initialization in a background thread and provide callbacks for the user to know when OTel is ready to be used.

In theory, it's ok but in practice, this has downsides, for example:
* Some features to be installed require to be in the main thread, which is fine, we can always post an execution to the main thread.
* For tracing instrumentation, the SDK must be init. as soon as possible, and we can only guarantee that if the SDK is init in the main thread, executing all the code sync.
* For crash reporting, it's also important that the SDK is init. as soon as possible, for example, to capture errors during app start.

Maybe an option is to init. the "core" parts of the SDK in the main thread and defer everything else to a worker.

Another option is just to defer the blocking bits to a worker (preferred IMO), and await the worker whenever needed even if that blocks for a tiny bit, an example of that is the storage init., it's gonna be init. in a worker, but if a trace or error has to be cached in storage and the worker is still ongoing, it's ok to `await` until it's finished.

For the NTP bits, I don't know any other solution than an NTP sync, there are [libs](https://github.com/instacart/truetime-android) that do that async instead, that doesn't need to be a blocking call, the very first few events before the NTP sync is finished might be off, but that's fine if all the following events are accurate.

Another option is doing a clock drift on the server side, for example, event A has 12:00 as a time when calling the API to send the event, you also send the current timestamp (request time), let's say, 12:10,  the server compares with its current timestamp and do a clock drift, if the server is 12:15 when it receives the request, and the request was done at 12:10 device time, the server knows that the device is 5 minutes late, so the event that was at 12:00 is 12:05.
This solution would force vendors to implement clock drift, so probably not ideal.

**Comment by bidetofevil** [comment]
Another option is to offer async init as an option (but not the default) with the known issues that that may bring. Having the SDK ready and instrumenting after the init method is invoked is powerful and gives us some guarantees about when telemetry is expected to roll in, but not everyone requires that level of guarantee, so it may make sense to offer that as an option for those who don't want disk reads of simply don't want to the app to be blocked on startup.

**Comment by LikeTheSalad** [comment]
Thanks @bidetofevil - I agree, it might be useful to provide that as an option just in case some projects would prefer losing some data over making reading disk operations or adding any kind of overhead during app launch 👍 

So to summarize, based on what's been discussed in this thread, it seems like this would be the overall conclusion:

* By default, we should keep the agent's initialization work in the main thread, except for the bits related to making network requests and/or writing disk operations (which there aren't any atm afaik, but I think we should still mention it for clarification purposes), in other words, doing a best effort to keep the initialization in the main thread while optimizing it whenever possible to avoid affecting the UI performance.
* We might want to provide an async option in the future for doing the whole initialization in a worker thread, which will provide a callback to let users know when the initialization is finished so they can start sending telemetry afterward. We should mention somewhere in the docs that, by using the async init option, there's a risk of losing some app launch related data.

**Comment by breedx-splk** [comment]
Related to #709, which defers some startup work to the background.

**Comment by fractalwrench** [comment]
Closing as `PreferencesService` is no longer a thing
